### PR TITLE
feat: split to create a library

### PR DESCRIPTION
**Pull request by jayvdb**
Not the cleanest split; happy to revise.
And also happy for this to not be published until it goes through a few more revisions to clean it up to produce a sensible API. (e.g. MainError probably doesnt belong in the lib)

### CODE REVIEW

**Comment by jayvdb** [comment]
@KSXGitHub , do you feel this might be possible ?

**Comment by jayvdb** [comment]
fwiw, I have developed a few other `pnpm` invocation wrappers which I would like to find a open source home for.  See https://github.com/simonbuchan/pnpm-extra/issues/2 for one of them.

**Comment by KSXGitHub** [comment]
> do you feel this might be possible ?

Technically possible. I have created plenty of Rust projects (not related to pnpm) that is both a library crate and a binary crate. All you have to do is to define both `[lib]` and `[[bin]]` in `Cargo.toml`.

**Comment by jayvdb** [comment]
> All you have to do is to define both [lib] and [[bin]] in Cargo.toml

I believe this isnt necessary unless there are attributes needed to describe them.

i.e.
having both src/lib.rs and src/main.rs is sufficient...?

`cargo run` and `cargo build` do the appropriate things afaics

File: src/utils.rs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+use std::io::ErrorKind;
+use std::num::NonZeroI32;
+use std::path::Path;
+use std::process::{Command, Stdio};
+
+use pipe_trait::Pipe;
+
+use crate::error::{MainError, PnError};
+use crate::shell_quoted::ShellQuoted;
+use crate::NodeManifest;
```
Comment (KSXGitHub) [inline src/utils.rs:13]: Let's not change how imports are organized.

```rust
use clap::Parser;
use cli::Cli;
use error::{MainError, PnError};
use indexmap::IndexMap;
use pipe_trait::Pipe;
use serde::{Deserialize, Serialize};
use shell_quoted::ShellQuoted;
use std::{
    env,
    ffi::OsString,
    fs::File,
    io::{self, ErrorKind, Write},
    num::NonZeroI32,
    io::{self, Write},
    path::Path,
    process::{exit, Command, Stdio},
    process::exit,
};
use yansi::Color::{Black, Red};
```

Comment (jayvdb) [reply src/utils.rs:13]: I was using the `group_imports = "StdExternalCrate"` setting of rustfmt nightly

Comment (KSXGitHub) [reply src/utils.rs:13]: Let's not change default.

**Comment by zkochan** [comment]
Why is a lib needed? Maintaining a library along with the CLI creates more work for us.

**Comment by jayvdb** [comment]
See earlier comment https://github.com/pnpm/pn/pull/62#issuecomment-2364757109

I would like to use `pn` in my Rust project.

**Review by zkochan — APPROVED** [review]

**Review by KSXGitHub — APPROVED** [review]

**Comment by jayvdb** [comment]
Thanks for approving.
I should add that I am happy to use `git = ".."` in my deps, so no need to do a release, and also I am expecting that the lib is very unstable.

It would be a good idea to follow this up with adding a header to the lib.rs to indicate that it is unstable, e.g. https://github.com/simonbuchan/pnpm-extra/blob/main/src/lib.rs , and maybe even hide it from the rust docs , e.g. https://github.com/Electron100/butane/blob/master/butane_cli/src/lib.rs

**Pull request merged**